cherab/jet/spectroscopy/kt3/kt3.py

```diff
@@ -0,0 +1,405 @@
+    KT3A spectrometer and t3bf for KT3B spectrometer. The spot position can be further corrected using the data from
+    "/pulse/#pulse/ppf/signal/ameigs/kt3s/shfm", which gives the radial shift of the spot due to shot interference.
+
+
+    It was also discovered, that sometimes the cherab sightlines can be obscured by the JET vessem models (probably
+    due to a uncertainty in the origin position) A shift to the sightline origin can be applied to remove the blocking.
+    The constant KT3_INNER_THRESHOLD is numerically obtained position of the bocking edge and can be recalculated using
+    the determine_inner_blockpoint function.
+
+    :param str spectrometer: Name of the spectrometer
+    :param int pulse_number: JET pulse number
+    :param float min_wavelength: Minimum observed wavelength to be used instead of the value provided in ppfs.
+    :param float max_wavelength: Maximum observed wavelength to be used instead of the value provided in ppfs.
+    :param int spectral_bins: Number of spectral bins
+    :param bool spot_position_correction: If True, the radial spot position correction will be applied. If true, the 
+      pulse_time parameter has to be passed.
+    :param float pulse_time: Shot time to apply the radial spot position for.
+    :param str time_selection: Selects the method the correction time will be selected with. If "precise", then the
+      pulse_time value is matched exactly to the time values in the kt3s/shfm signal. If "nearest", then the nearest
+      value to the value passed as pulse_time is provided (default "nearest").
+    :param int pixel_samples: Number of sample rays to be used for every sight-line
+    :param bool prevent_blocking: If true, the position of the origin is shifted to avoid blocking of the inner sight-lines (default True).
+    :param Point2D inner_threshold: The point setting the lowest (min(z)) outer most (max(r)) point of the limiting structure. Sight-line origins
+      will be moved otwards to avoid blocking.
+    :parameter parent: The parent of the returned FibreOpticGroup, (optional)
+    :return: FibreOpticGroup with kt3 sight-lines
+    """
+
+    # get KT3A/KT3B data
+    if spectrometer.lower() in ("a", "kt3a"):
+        leaf_origin = "/pulse/{0:d}/ppf/signal/jetppf/t3af/rzor".format(int(pulse_number))
+        leaf_spot_tile5 = "/pulse/{0:d}/ppf/signal/jetppf/t3af/rzdv".format(int(pulse_number))
+        leaf_wave = "/pulse/{0:d}/ppf/signal/jetppf/t3af/wave".format(int(pulse_number))
+        spectrometer = "KT3A"
+    elif spectrometer.lower() in ("b", "kt3b"):
+        leaf_origin = "/pulse/{0:d}/ppf/signal/jetppf/t3bf/rzor".format(int(pulse_number))
+        leaf_spot_tile5 = "/pulse/{0:d}/ppf/signal/jetppf/t3bf/rzdv".format(int(pulse_number))
+        leaf_wave = "/pulse/{0:d}/ppf/signal/jetppf/t3bf/wave".format(int(pulse_number))
+        spectrometer = "KT3B"
```

Comment:
Thanks for pointing this out @vsnever . I didn't notice because the shot ensemble I'm working with. I will contact Andy Meigs to discus with him what could be the best way forward, but for myself:

1. Would cause a lot of problems to a lot of people and as you said it would not work remotely. I think this is not an option.
2. This would be the best solution in my opinion. Or getting the data from other files, we will see.
3. Having a default data would be a last resort solution.

I will inform here when I get some information.

Reply:
I'm currently working on the 3rd solution by obtaining the values of `spots_rz` using `kt3a(b)_los_siw_new()` for the entire set of ILW pulses. So far I have found that `spots_rz` depends on the grating, which varies from pulse to pulse, but can only take on 3 values: 300, 1200 and 2400, while for the same grating, the `spots_rz` changes very rarely.

Reply:
Lets wait for Andy Meigs before we jump to reverse engineering.